### PR TITLE
feature/recommendations

### PR DESCRIPTION
**Pull request by xidealo**

### CODE REVIEW

File: app/src/main/java/com/bunbeauty/papakarlo/feature/auth/LoginFragment.kt [hunk 1]
```diff
@@ -49,7 +49,7 @@ class LoginFragment : BaseComposeFragment<Login.State, Login.Action, Login.Event
 
     private val args: LoginFragmentArgs by navArgs()
 
-    private val LogoBoxHeightLimit: Dp = 200.dp
+    private val logoBoxHeightLimit: Dp = 200.dp
```
Comment (MaxAstin) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/auth/LoginFragment.kt:52]: Let's change it back according Compose [code style](https://github.com/androidx/androidx/blob/androidx-main/compose/docs/compose-api-guidelines.md#singletons-constants-sealed-class-and-enum-class-values)

File: app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt [hunk 2]
```diff
@@ -38,145 +38,236 @@ import com.bunbeauty.papakarlo.common.ui.screen.ErrorScreen
...
 import com.bunbeauty.shared.presentation.consumercart.ConsumerCartData
-import com.bunbeauty.shared.presentation.consumercart.ConsumerCartEvent
-import com.bunbeauty.shared.presentation.consumercart.ConsumerCartUIState
+import com.bunbeauty.shared.presentation.consumercart.ConsumerCartState
 import com.bunbeauty.shared.presentation.consumercart.ConsumerCartViewModel
 import org.koin.androidx.viewmodel.ext.android.viewModel
 
-class ConsumerCartFragment : BaseFragmentWithSharedViewModel(R.layout.layout_compose) {
+class ConsumerCartFragment :
+    BaseComposeFragment<ConsumerCartState.State, ConsumerCartState.Action, ConsumerCartState.Event>() {
```
Comment (MaxAstin) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt:57]: Lets rename `ConsumerCartState` to `ConsumerCart`

File: app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt [hunk 3]
```diff
@@ -38,145 +38,236 @@ import com.bunbeauty.papakarlo.common.ui.screen.ErrorScreen
...
-            when (consumerCartState) {
-                ConsumerCartUIState.ConsumerCartState.Loading -> LoadingScreen()
-                is ConsumerCartUIState.ConsumerCartState.Success -> ConsumerCartSuccessScreen(
-                    consumerCartData = consumerCartState.data,
-                    addProductToCartClicked = addProductToCartClicked,
-                    removeProductFromCartClicked = removeProductFromCartClicked,
-                    onProductClicked = onProductClicked,
-                    onCreateOrderClicked = onCreateOrderClicked
-                )
+            // check if change state to sealed class
```
Comment (MaxAstin) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt:81]: Lets remove reminder comments

File: app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt [hunk 4]
```diff
@@ -38,145 +38,236 @@ import com.bunbeauty.papakarlo.common.ui.screen.ErrorScreen
...
+            Crossfade(targetState = state.screenState, label = "ConsumerCart") { screenState ->
+                when (screenState) {
+                    ConsumerCartState.ScreenState.LOADING -> LoadingScreen()
+                    ConsumerCartState.ScreenState.SUCCESS -> {
+                        // Warning(Check!)
```
Comment (MaxAstin) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt:86]: Lets remove reminder comments

File: app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt [hunk 5]
```diff
@@ -38,145 +38,236 @@ import com.bunbeauty.papakarlo.common.ui.screen.ErrorScreen
...
-                .background(FoodDeliveryTheme.colors.mainColors.background)
         ) {
             Box(modifier = Modifier.weight(1f)) {
-                LazyColumn(
+                LazyVerticalGrid(
                     modifier = Modifier.fillMaxSize(),
-                    contentPadding = PaddingValues(FoodDeliveryTheme.dimensions.mediumSpace)
+                    contentPadding = PaddingValues(vertical = FoodDeliveryTheme.dimensions.mediumSpace),
+                    columns = GridCells.Fixed(2),
+                    horizontalArrangement = Arrangement.Absolute.spacedBy(8.dp)
```
Comment (MaxAstin) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt:166]: `Absolute` is extra, lets change to `Arrangement.spacedBy(8.dp)`

File: app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt [hunk 6]
```diff
@@ -38,145 +38,236 @@ import com.bunbeauty.papakarlo.common.ui.screen.ErrorScreen
...
+                                        )
+                                    )
+                                }
+                            )
+                            if (index != consumerCartData.cartProductList.lastIndex) {
+                                Divider(
+                                    modifier = Modifier
+                                        .padding(horizontal = 16.dp),
+                                    thickness = 1.dp,
+                                    color = FoodDeliveryTheme.colors.mainColors.background
```
Comment (MaxAstin) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt:221]: Lets add new color `FoodDeliveryTheme.colors.mainColors.stroke` according design

File: app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt [hunk 7]
```diff
@@ -38,145 +38,236 @@ import com.bunbeauty.papakarlo.common.ui.screen.ErrorScreen
...
                         )
                     }
-                    itemsIndexed(consumerCartData.cartProductList) { i, cartProductItemModel ->
-                        CartProductItem(
-                            modifier = Modifier.padding(
-                                top = FoodDeliveryTheme.dimensions.getItemSpaceByIndex(i)
+
+                    itemsIndexed(
+                        items = consumerCartData.cartProductList,
+                        key = { index, cartProductItem -> cartProductItem.uuid },
```
Comment (MaxAstin) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt:188]: U2U replace unused params with `_`

File: app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt [hunk 8]
```diff
@@ -38,145 +38,236 @@ import com.bunbeauty.papakarlo.common.ui.screen.ErrorScreen
...
-                            onCountIncreased = {
-                                addProductToCartClicked(cartProductItemModel.menuProductUuid)
-                            },
-                            onCountDecreased = {
-                                removeProductFromCartClicked(cartProductItemModel.menuProductUuid)
+                            style = FoodDeliveryTheme.typography.titleMedium.medium,
+                            color = FoodDeliveryTheme.colors.mainColors.onSurface
+                        )
+                    }
+                    itemsIndexed(
```
Comment (MaxAstin) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt:241]: Lets use `items() { ... }`

File: app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt [hunk 9]
```diff
@@ -269,29 +334,37 @@ class ConsumerCartFragment : BaseFragmentWithSharedViewModel(R.layout.layout_com
             menuProductUuid = ""
         )
 
+        fun getMenuProductItem(uuid: String) =
+            com.bunbeauty.shared.presentation.menu.MenuProductItem(
```
Comment (MaxAstin) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt:338]: Lets as alias:
`import com.bunbeauty.shared.presentation.menu.MenuProductItem as MenuProductItemModel`

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/data/mapper/RecommendationMapper.kt [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+                uuid = recommendationProductWithMenuProductEntity.uuid,
+                name = recommendationProductWithMenuProductEntity.name,
+                newPrice = recommendationProductWithMenuProductEntity.newPrice,
+                oldPrice = recommendationProductWithMenuProductEntity.oldPrice,
+                utils = recommendationProductWithMenuProductEntity.utils,
+                nutrition = recommendationProductWithMenuProductEntity.nutrition,
+                description = recommendationProductWithMenuProductEntity.description,
+                comboDescription = recommendationProductWithMenuProductEntity.comboDescription,
+                photoLink = recommendationProductWithMenuProductEntity.photoLink,
+                categoryList = emptyList(),
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/data/mapper/RecommendationMapper.kt:39]: We need this list for filtration

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/domain/feature/cart/GetRecommendationsUseCase.kt [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+
+import com.bunbeauty.shared.data.repository.RecommendationRepository
+import com.bunbeauty.shared.domain.model.RecommendationProduct
+import com.bunbeauty.shared.domain.model.product.MenuProduct
+
+class GetRecommendationsUseCase(
+    private val recommendationRepository: RecommendationRepository,
+) {
+    suspend operator fun invoke(): List<RecommendationProduct> {
+        return recommendationRepository.getRecommendations()
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/domain/feature/cart/GetRecommendationsUseCase.kt:11]: Let's add a little bit logic, need filtration

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/consumercart/ConsumerCartViewModel.kt [hunk 12]
```diff
@@ -4,72 +4,120 @@ import com.bunbeauty.shared.Constants.PERCENT
...
+    override fun handleAction(action: ConsumerCartState.Action) {
+
+        when (action) {
+            is ConsumerCartState.Action.AddProductToCartClick -> onAddCardProductClicked(
+                menuProductUuid = action.menuProductUuid
+            )
+
+            ConsumerCartState.Action.BackClick -> navigateBack()
+            is ConsumerCartState.Action.ConsumeEvents -> consumeEvents(action.eventList)
+            ConsumerCartState.Action.Init -> init()
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/consumercart/ConsumerCartViewModel.kt:51]: We can do it inside `init { ... }` block without `Action`

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/consumercart/ConsumerCartViewModel.kt [hunk 13]
```diff
@@ -4,72 +4,120 @@ import com.bunbeauty.shared.Constants.PERCENT
...
-    fun getConsumerCart() {
+    override fun handleAction(action: ConsumerCartState.Action) {
+
+        when (action) {
+            is ConsumerCartState.Action.AddProductToCartClick -> onAddCardProductClicked(
+                menuProductUuid = action.menuProductUuid
+            )
+
+            ConsumerCartState.Action.BackClick -> navigateBack()
+            is ConsumerCartState.Action.ConsumeEvents -> consumeEvents(action.eventList)
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/consumercart/ConsumerCartViewModel.kt:50]: We already consume `Events` in Base class

**Review by MaxAstin — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt (hunk 2 above)
Comment (xidealo) [reply app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt:57]: We have same name for another model
Can i rename ConsumerCart in domain to ConsumerCartDomain?

File: app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt (hunk 3 above)
Comment (xidealo) [reply app/src/main/java/com/bunbeauty/papakarlo/feature/consumercart/ConsumerCartFragment.kt:81]: Will change it?

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/domain/feature/cart/GetRecommendationsUseCase.kt [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+                if (!hasSameCategory) {
+                    add(recommendationProduct)
+                }
+            }
+        }
+
+        return if (recommendationProductListNoneCategory.size < 2) {
+            buildList {
+                addAll(recommendationProductListNoneCategory)
+                recommendationProductList.forEach { recommendationProduct ->
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/domain/feature/cart/GetRecommendationsUseCase.kt:41]: 1. Lets move this code to separate function
2. Add condition that recommendation not in `recommendationProductListNoneCategory` (not added yet)

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/domain/feature/cart/GetRecommendationsUseCase.kt [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+                    if (cartProductList.none { it.product.uuid == recommendationProduct.menuProduct.uuid }) {
+                        add(recommendationProduct)
+                    }
+                }
+            }.take(recommendation.maxVisibleCount)
+        } else {
+            recommendationProductListNoneCategory.take(recommendation.maxVisibleCount)
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/domain/feature/cart/GetRecommendationsUseCase.kt:48]: Let's get the list and then call `take(...)` and return the result on the next line

File: shared/src/commonTest/kotlin/com/bunbeauty/domain/feature/recommendation/GetRecommendationsUseCaseTest.kt [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+        coEvery { recommendationRepository.getRecommendations() } returns getRecommendationProductList()
+
+        coEvery { cartProductInteractor.getCartProductList() } returns listOf(
+            getCartProduct(
+                uuid = "1",
+                count = 1,
+                menuProduct = getMenuProduct("1"),
+            )
+        )
+        assertTrue(getRecommendationsUseCase().isEmpty())
```
Comment (MaxAstin) [inline shared/src/commonTest/kotlin/com/bunbeauty/domain/feature/recommendation/GetRecommendationsUseCaseTest.kt:56]: Lets check list content, not just size

**Review by MaxAstin — CHANGES_REQUESTED** [review]

File: shared/src/commonTest/kotlin/com/bunbeauty/domain/feature/recommendation/GetRecommendationsUseCaseTest.kt [hunk 17]
```diff
@@ -0,0 +1,388 @@
...
+                    ),
+                )
+            )
+            assertTrue(getRecommendationsUseCase().size == 3)
+        }
+
+    @Test
+    fun `return recommendations when cartProducts has same category but not all and many cartProducts`() =
+        runTest {
+            val categoryList = listOf(
```
Comment (MaxAstin) [inline shared/src/commonTest/kotlin/com/bunbeauty/domain/feature/recommendation/GetRecommendationsUseCaseTest.kt:288]: Lets remove it if it's unused

File: shared/src/commonTest/kotlin/com/bunbeauty/domain/feature/recommendation/GetRecommendationsUseCaseTest.kt [hunk 18]
```diff
@@ -0,0 +1,388 @@
...
+                getCategoryProduct(
+                    uuid = "1",
+                )
+            )
+            coEvery { recommendationRepository.getRecommendations() } returns getRecommendationProductList(
+                maxVisibleCount = 4,
+                recommendationProductList = listOf(
+                    RecommendationProduct(
+                        uuid = "1",
+                        menuProduct = getMenuProduct(
```
Comment (MaxAstin) [inline shared/src/commonTest/kotlin/com/bunbeauty/domain/feature/recommendation/GetRecommendationsUseCaseTest.kt:298]: 1. Lets create `fun getMenuProductWithCategory(categoryUuid = ...)`
2. Lets write it on one line if we only have one param to make it more compact
3. Lets create similar fun for `CartProduct`

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/domain/feature/cart/GetRecommendationsUseCase.kt [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+        cartProductList: List<CartProduct>,
+    ) = recommendationProduct.menuProduct.categoryList.any { recommendationProductCategory ->
+        hasRecommendationCategoriesInCartProduct(cartProductList, recommendationProductCategory)
+    }
+
+    private fun hasRecommendationCategoriesInCartProduct(
+        cartProductList: List<CartProduct>,
+        recommendationProductCategory: Category,
+    ) = cartProductList
+        .any { cartProduct ->
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/domain/feature/cart/GetRecommendationsUseCase.kt:64]: Change formatting to
```
 list.any { item ->
   ...
 }
```

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/domain/feature/cart/GetRecommendationsUseCase.kt [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+        val recommendationProductList =
+            recommendationsWithMaxCount?.recommendationProductList ?: return emptyList()
+
+        val recommendationProductListNoneCategory = getFilteredRecommendationsByCategories(
+            recommendationProductList = recommendationProductList,
+            cartProductList = cartProductList
+        )
+
+        return if (recommendationProductListNoneCategory.size < 2) {
+            recommendationProductListNoneCategory + getRecommendationsByCartAndFilteredRecommendations(
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/domain/feature/cart/GetRecommendationsUseCase.kt:27]: Please create a variable to make it more readable

**Review by MaxAstin — CHANGES_REQUESTED** [review]

File: shared/src/commonTest/kotlin/com/bunbeauty/domain/feature/recommendation/GetRecommendationsUseCaseTest.kt [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+                    RecommendationProduct(
+                        uuid = "1",
+                        menuProduct = getMenuProductWithCategory("1")
+                    )
+                )
+            )
+            coEvery { cartProductInteractor.getCartProductList() } returns listOf(
+                getCartProductWithCategory("1")
+            )
+            assertTrue(getRecommendationsUseCase().isEmpty())
```
Comment (MaxAstin) [inline shared/src/commonTest/kotlin/com/bunbeauty/domain/feature/recommendation/GetRecommendationsUseCaseTest.kt:86]: Why? It should or shouldn't return recommendation depending on `menuProductId`

Comment (MaxAstin) [reply shared/src/commonTest/kotlin/com/bunbeauty/domain/feature/recommendation/GetRecommendationsUseCaseTest.kt:86]: Let's add `menuProductId` param to have clearer tests

**Review by MaxAstin — CHANGES_REQUESTED** [review]

**Review by MaxAstin — APPROVED** [review]

**Pull request merged**